File: python/examples/01_mix_analysis.py
#!/usr/bin/env python3
"""
01_mix_analysis.py - Analyze Your Mix with the Tonn API

This example demonstrates how to analyze a mixed or mastered audio track
using the Tonn API's mix diagnosis feature (powered by Mix Check Studio).

What this does:
- Uploads your audio file to cloud storage
- Analyzes loudness, dynamic range, stereo field, and tonal balance
- Provides actionable feedback for improving your mix

Usage:
    python 01_mix_analysis.py <audio_file> <musical_style> [--is-master]

Example:
    python 01_mix_analysis.py my_track.wav POP --is-master
    python 01_mix_analysis.py demo.mp3 ROCK

Musical styles: ROCK, POP, ELECTRONIC, HIPHOP_GRIME, ACOUSTIC, ROCK_INDIE, etc.
"""

import sys
import os
import argparse

# Add the parent directory to path so we can import shared modules
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from shared import TonnClient


def print_diagnosis_results(diagnosis: dict) -> None:
    """Pretty-print the mix diagnosis results."""
    if not isinstance(diagnosis, dict):
        print("No valid diagnosis results found.")
        return

    print("\n" + "=" * 60)
    print("MIX DIAGNOSIS RESULTS")
    print("=" * 60)

    # Completion info
    print(f"\nCompletion Time: {diagnosis.get('completion_time', 'N/A')}")
    if diagnosis.get('error'):
        print(f"⚠️  Error: {diagnosis.get('info', 'Unknown error')}")
        return

    # Payload details
    payload = diagnosis.get("payload", {})
    if payload:
        print("\n--- Technical Details ---")
        
        # Key metrics
        metrics = [
            ("Bit Depth", payload.get("bit_depth")),
            ("Sample Rate", payload.get("sample_rate")),
            ("Integrated Loudness (LUFS)", payload.get("integrated_loudness_lufs")),
            ("Peak Loudness (dBFS)", payload.get("peak_loudness_dbfs")),
            ("Clipping", payload.get("clipping")),
            ("Stereo Field", payload.get("stereo_field")),
            ("Mono Compatible", payload.get("mono_compatible")),
            ("Phase Issues", payload.get("phase_issues")),
        ]
        
        for name, value in metrics:
            if value is not None:
                print(f"  {name}: {value}")
        
        # Mastering evaluation
        if payload.get("if_master_drc"):
            print(f"\n  Master DRC Evaluation: {payload.get('if_master_drc')}")
        if payload.get("if_master_loudness"):
            print(f"  Master Loudness Evaluation: {payload.get('if_master_loudness')}")

    # Summary
    summary = payload.get("summary", {})
    if summary:
        print("\n--- Recommendations ---")
        for idx, (key, value) in enumerate(summary.items(), 1):
            print(f"  {idx}. {value}")

    print("\n" + "=" * 60)


def main():
    parser = argparse.ArgumentParser(
        description="Analyze a mix/master using the Tonn API",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python 01_mix_analysis.py track.wav POP
  python 01_mix_analysis.py master.mp3 ROCK --is-master
  python 01_mix_analysis.py demo.flac ELECTRONIC

Musical Styles:
  ROCK, POP, ELECTRONIC, HIPHOP_GRIME, ACOUSTIC, ROCK_INDIE, 
  SINGER_SONGWRITER, JAZZ, CLASSICAL, and more.
        """
    )
    parser.add_argument("input_file", help="Path to audio file (.mp3, .wav, .flac)")
    parser.add_argument("musical_style", help="Musical style (e.g., ROCK, POP)")
    parser.add_argument("--is-master", action="store_true", 
                       help="Flag if the input is a mastered track")

    args = parser.parse_args()

    # Validate input file exists
    if not os.path.isfile(args.input_file):
        print(f"Error: File not found: {args.input_file}")
        sys.exit(1)

    # Initialize client
    client = TonnClient()

    # Upload the file
    print(f"\n📤 Uploading {os.path.basename(args.input_file)}...")
    readable_url = client.upload_local_file(args.input_file)
    if not readable_url:
        print("Failed to upload file. Exiting.")
        sys.exit(1)

    # Build the analysis payload
    payload = {
        "mixDiagnosisData": {
            "audioFileLocation": readable_url,
            "musicalStyle": args.musical_style.upper(),
            "isMaster": args.is_master
        }
    }

    # Run the analysis
    print(f"\n🔍 Analyzing your {'master' if args.is_master else 'mix'}...")
    response = client.post("/mixanalysis", payload)
    
    if not response:
        print("Analysis failed. Check your API key and try again.")
        sys.exit(1)

    # Display results
    if response.get("error"):
        print(f"❌ Error: {response.get('message', 'Unknown error')}")
        sys.exit(1)

    print(f"✅ {response.get('message', 'Analysis complete')}")
    
    diagnosis_results = response.get("mixDiagnosisResults")
    if diagnosis_results:
        print_diagnosis_results(diagnosis_results)
    else:
        print("No diagnosis results in response.")


if __name__ == "__main__":
    main()

File: nodejs/src/examples/01_mix_analysis.js
#!/usr/bin/env node
/**
 * 01_mix_analysis.js - Analyze Your Mix with the Tonn API
 *
 * This example demonstrates how to analyze a mixed or mastered audio track
 * using the Tonn API's mix diagnosis feature (powered by Mix Check Studio).
 *
 * Usage:
 *   node 01_mix_analysis.js <audio_file> <musical_style> [--is-master]
 *
 * Example:
 *   node 01_mix_analysis.js my_track.wav POP --is-master
 *   node 01_mix_analysis.js demo.mp3 ROCK
 */

import { existsSync } from 'fs';
import { TonnClient } from '../shared/client.js';

/**
 * Print diagnosis results in a readable format.
 */
function printDiagnosisResults(diagnosis) {
  if (!diagnosis || typeof diagnosis !== 'object') {
    console.log('No valid diagnosis results found.');
    return;
  }

  console.log('\n' + '='.repeat(60));
  console.log('MIX DIAGNOSIS RESULTS');
  console.log('='.repeat(60));

  console.log(`\nCompletion Time: ${diagnosis.completion_time || 'N/A'}`);

  if (diagnosis.error) {
    console.log(`⚠️  Error: ${diagnosis.info || 'Unknown error'}`);
    return;
  }

  const payload = diagnosis.payload || {};
  if (Object.keys(payload).length > 0) {
    console.log('\n--- Technical Details ---');

    const metrics = [
      ['Bit Depth', payload.bit_depth],
      ['Sample Rate', payload.sample_rate],
      ['Integrated Loudness (LUFS)', payload.integrated_loudness_lufs?.toFixed(1)],
      ['Peak Loudness (dBFS)', payload.peak_loudness_dbfs],
      ['Clipping', payload.clipping],
      ['Stereo Field', payload.stereo_field],
      ['Mono Compatible', payload.mono_compatible],
      ['Phase Issues', payload.phase_issues]
    ];

    for (const [name, value] of metrics) {
      if (value !== undefined && value !== null) {
        console.log(`  ${name}: ${value}`);
      }
    }

    if (payload.if_master_drc) {
      console.log(`\n  Master DRC Evaluation: ${payload.if_master_drc}`);
    }
    if (payload.if_master_loudness) {
      console.log(`  Master Loudness Evaluation: ${payload.if_master_loudness}`);
    }
  }

  const summary = payload.summary || {};
  if (Object.keys(summary).length > 0) {
    console.log('\n--- Recommendations ---');
    let idx = 1;
    for (const value of Object.values(summary)) {
      console.log(`  ${idx}. ${value}`);
      idx++;
    }
  }

  console.log('\n' + '='.repeat(60));
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length < 2 || args.includes('--help') || args.includes('-h')) {
    console.log(`
Usage: node 01_mix_analysis.js <audio_file> <musical_style> [--is-master]

Arguments:
  audio_file     Path to audio file (.mp3, .wav, .flac)
  musical_style  Musical style (e.g., ROCK, POP, ELECTRONIC)
  --is-master    Flag if the input is a mastered track

Examples:
  node 01_mix_analysis.js track.wav POP
  node 01_mix_analysis.js master.mp3 ROCK --is-master

Musical Styles:
  ROCK, POP, ELECTRONIC, HIPHOP_GRIME, ACOUSTIC, ROCK_INDIE,
  SINGER_SONGWRITER, JAZZ, CLASSICAL, and more.
    `);
    process.exit(args.includes('--help') ? 0 : 1);
  }

  const inputFile = args[0];
  const musicalStyle = args[1].toUpperCase();
  const isMaster = args.includes('--is-master');

  // Validate input file
  if (!existsSync(inputFile)) {
    console.error(`Error: File not found: ${inputFile}`);
    process.exit(1);
  }

  const client = new TonnClient();

  // Upload the file
  console.log(`\n📤 Uploading ${inputFile}...`);
  const readableUrl = await client.uploadLocalFile(inputFile);

  if (!readableUrl) {
    console.error('Failed to upload file. Exiting.');
    process.exit(1);
  }

  // Build the analysis payload
  const payload = {
    mixDiagnosisData: {
      audioFileLocation: readableUrl,
      musicalStyle,
      isMaster
    }
  };

  // Run the analysis
  console.log(`\n🔍 Analyzing your ${isMaster ? 'master' : 'mix'}...`);
  const response = await client.post('/mixanalysis', payload);

  if (!response) {
    console.error('Analysis failed. Check your API key and try again.');
    process.exit(1);
  }

  if (response.error) {
    console.error(`❌ Error: ${response.message || 'Unknown error'}`);
    process.exit(1);
  }

  console.log(`✅ ${response.message || 'Analysis complete'}`);

  const diagnosisResults = response.mixDiagnosisResults;
  if (diagnosisResults) {
    printDiagnosisResults(diagnosisResults);
  } else {
    console.log('No diagnosis results in response.');
  }
}

main().catch(console.error);

